DOCS-11189: Metadata Assistant Feature (#244)
mulesoft/docs-studio: modules/ROOT/pages/metadata-propagation-between-flows.adoc

> = Metadata Propagation Between Flows

Reviewer: After reading through and editing the existing text, I finally came to understand these points:

- In Anypoint Studio 7.10, you can add a Flow Reference component to one flow that automatically propagates the metadata in that flow (_actual_ metadata) to all other flows it references, as _custom_ (or _expected_) metadata, so that you don't have to add the actual metadata to the referenced flows manually.
- This enables you to scaffold your flows using an inheritance model that automatically adds or updates custom metadata throughout your app.
- However, any subsequent change to actual metadata automatically overwrites the previously propagated custom/expected metadata; if you do not want this to happen, you must remove the component (?) and manually customize the metadata in referenced flows.

Is this correct?
The current use of synonymous names for types of flows and metadata confused me, so I think some term definition in the context of the intro narrative might help.
A graphic might help, too.

Reviewer: I've left in the original edits, just so you can see how I experienced the text in a realtime reading flow.
Please let me know if chatting about this would be easier.  :-)

> Anypoint Studio 7.10 infers the metadata of your current flow so that, when you use a Flow Reference component to reference another flow, you can set the metadata from your current flow as custom metadata for your referenced flow. By automatically detecting the actual metadata of your existing flow, Studio enables you to propagate the metadata in your flow to the flow you are referencing.

Reviewer: I wonder if we need both sentences, since the following two phrases are largely repetitive, just using different terms for the same thing:

"...when you use a Flow Reference component to reference another flow, you can **set** the metadata from your **current flow** as custom metadata for **your referenced flow.**"
and
"...automatically detecting the actual metadata of your **existing flow**, Studio enables you to **propagate** the metadata in **your flow** to **the flow you are referencing**."

Reviewer: RE:

set vs. propagate
current flow, existing flow, your flow
your referenced flow, the flow you are referencing

Also note that the first refers to "custom metadata"  but the second does not, and the first refers to "infer" and the second to to "automatic" detection.

> By leveraging Studio to define the metadata you can:

Reviewer: Which metadata: current, custom, actual? "Metadata" is referred to in many different ways earlier.
Also comma:
```suggestion
By leveraging Studio to define the metadata, you can:
```

> * Automatically propagate the metadata of your flow to a referenced flow without custom metadata configured.

Reviewer: ```suggestion
* Automatically propagate the metadata of your current flow to a referenced flow without configuring custom metadata.
```
This is confusing when considered alongside of the previous text about setting current metadata as **custom** metadata. Maybe we should say "automatically propagate rather than adding by configuration"?

> * Easily define the metadata of each referenced flow within your application.

Reviewer: This seems to be the focus of all this text: automatically propagating the metadata of your current flow to other flows referenced by it.

I suggest simplifying and reducing this text to something like...:

"Using the Flow Reference component in Anypoint Studio 7.10 enables you to automatically rather than manually propagate metadata in your current flow to each flow referenced by it."

> * Set custom metadata in a new flow without manually defining it.

Reviewer: See prev. suggested wording: it contains all three bullets, as well as the intro sentences.

> When you use Studio to scaffold your flows based on a RAML specification, Studio infers the metadata of your flow based on your RAML specification's types. However, note that Union types are not supported

Reviewer: Note that this sentence introduces the term "scaffold your flows": should we use this term earlier to define the relationships among current and referenced flows?

Reviewer: This text is specific to RAML specification and the previous is not? In other words, does this sentence refer to a specific situation that is is subset of that described previously?
If so, consider...:
```suggestion
Note that when you use Studio to scaffold your flows based on a RAML specification, Studio infers the metadata of those flows based on your RAML specification's types. However, note that Union types are not supported.
```

Reviewer: If Union types are not supported, what happens to a flow that uses Union types?

> Studio converts your inferred metadata as custom metadata for the flow you are referencing.

Reviewer: This sentence repeats all of the previous text and is unnecessary, except to return the audience to the main focus from the sidetrack of the RAML situation. If this is the case, perhaps the RAML text should follow the example?

> For example, assume a flow named `flow-1` that uses a Set Payload component to create a JSON payload:

Reviewer: ```suggestion
For example, consider a flow named `flow-1` that uses a Set Payload component to create a JSON payload:
```

.Actual Metadata Sample
[source,JSON,linenums]
--
{
  "value": {
    "ID": 1,
    "departureDate": "2016/12/20",
    "emptySeats": 200,
    "plane": {
      "type": "Boeing 747",
      "totalSeats": 400
    }
  }
}
--

> When you add a Flow Reference component to a second flow, `flow-2`, Studio detects the current metadata as the actual metadata that the referenced flow will receive and automatically assigns it as custom metadata in `flow-2` if the flow does not already have custom metadata declared for it.

Reviewer: So, the user must ADD a Flow Reference component to a flow? That should be specified earlier.

Reviewer: Again, I wonder about the use of "current," "actual," and "custom" metadata, all in one sentence.
I think maybe these terms need defining within the intro narrative.

Reviewer: "will receive" > "receives"

> Because the metadata in `flow-2` is set once, it can't update automatically if you modify the metadata in `flow-1`. For this example, if you choose to modify the actual metadata sample for the Set Payload component in `flow-1`, you must update the custom metadata in `flow-2`.

Reviewer: "Because...set once" refers to an assumption that the audience already knows this, so I think we want to tweak it to sound more explanatory:
```suggestion
Note that automatic metadata propagation from one flow to another can occur only one time. If you later modify the metadata in `flow-1`, you must manually update `flow-2`.
```

Reviewer: ...or is this true?
Can't automatic metadata propagation occur more than once...assuming that you want to overwrite the custom metadata?

> Consider now that `flow-2` has custom metadata configured already using the following JSON example:

Reviewer: Is this an example of the text at Line 35? If so, consider tying the two paras more closely, something like...

Automatic metadata propagation from one flow to another can occur only once. For example, after propagating metadata from `flow 1` to `flow 2`,  if you later decide to modify the metadata in `flow 1`, you must either manually modify `flow 2` as well, or allow this modification to overwrite the original propagated metadata?

.Expected Metadata Sample
[source,JSON,linenums]
--
{
  "value": {
    "ID": 1,
    "code": "GQ574",
    "price": 399,
    "departureDate": "2016/12/20",
    "origin": "ORD",
    "destination": "SFO",
    "emptySeats": 200,
    "plane": {
      "type": "Boeing 747",
      "totalSeats": 400
    }
  }
}
--

> Studio considers The custom metadata of `flow-2` as the _expected_ metadata for the referenced flow, and the metadata set by the Set Payload component in `flow-1` as its _actual_ metadata. You can choose to either overwrite the expected metadata with the actual metadata, or customize it so that you choose what metadata you want to overwrite, and which metadata you want to preserve.

Reviewer: lowercase "the"

Reviewer: Once again, I get tangled up in terminology: here we add "expected" and "actual" metadata to existing/current metadata and custom metadata.
